lib/src/closure/closure_codegen.dart

```diff
@@ -0,0 +1,144 @@
+    var namedArgs = <String, ClosureType>{};
+    for (var param in e.parameters) {
+      var t = _closureTypeForDartType(param.type);
+      switch (param.parameterKind) {
+        case ParameterKind.NAMED:
+          namedArgs[param.name] = t.orUndefined();
+          break;
+        case ParameterKind.POSITIONAL:
+          paramTypes[param.name] = t.toOptional();
+          break;
+        case ParameterKind.REQUIRED:
+          paramTypes[param.name] = t;
+          break;
+      }
+    }
+    if (namedArgs.isNotEmpty) {
+      paramTypes[namedArgsMapName] = new ClosureType.record(namedArgs).toOptional();
+    }
+
+    bool isConstructor = e is ConstructorElement && !e.isFactory && !generatesES6Classes;
+    bool isFactory = e is ConstructorElement && e.isFactory;
+    enclosingType() => _closureTypeForClass(e.enclosingElement);
+    
+    var returnType =
+        isFactory ? enclosingType() : _closureTypeForDartType(e.returnType);
+
+    return new ClosureAnnotation(
+        isConstructor: isConstructor,
+        isStruct: isConstructor,
+        isOverride: e.isOverride,
+        thisType: e.enclosingElement is ClassElement && !e.isStatic 
+            && !generatesES6Classes ?
+            enclosingType() : null,
+        // Note: Dart and Closure privacy are not compatible: don't set `isPrivate: e.isPrivate`.
+        paramTypes: paramTypes,
+        returnType: returnType);
+  }
+
+  Map<DartType, ClosureType> __commonTypes;
+  Map<DartType, ClosureType> get _commonTypes {
+    if (__commonTypes == null) {
+      var numberType = new ClosureType.number().toNullable();
+      __commonTypes = {
+        types.intType: numberType,
+        types.numType: numberType,
+        types.doubleType: numberType,
+        types.boolType: new ClosureType.boolean().toNullable(),
+        types.stringType: new ClosureType.string(),
+      };
+    }
+    return __commonTypes;
+  }
+    
+  ClosureType _closureTypeForClass(ClassElement classElement, [DartType type]) {
+    ClosureType closureType = _commonTypes[type];
+    if (closureType != null) return closureType;
+
+    var fullName = _getFullName(classElement);
+    switch (fullName) {
+      // TODO(ochafik): Test DartTypes directly if possible.
+      case "dart.js.JsArray":
+        return new ClosureType.array(
+            type is InterfaceType && type.typeArguments.length == 1
+            ? _closureTypeForDartType(type.typeArguments.single) : null);
+      case "dart.js.JsObject":
+        return new ClosureType.map();
+      case "dart.js.JsFunction":
+        return new ClosureType.function();
+      default:
+        return new ClosureType.type(getQualifiedName(classElement));
+    }
+  }
+
+  ClosureType _closureTypeForDartType(DartType type) {
+    // TODO(ochafik): Work out the practical difference between all (*) and unknown (?).
+    if (type == null) return new ClosureType.unknown();
+    if (type.isDynamic) return new ClosureType.all();
```

Comment:
Hehe, so I've [played a bit with function params](https://gist.github.com/ochafik/d8f701557f5bcab68940) and downcasts in Closure, following @leafpetersen's comments, and looks like we probably want to map pretty much everything weirdly-typed (dynamic, bottom) to unknown `?` (see [doc](https://developers.google.com/closure/compiler/docs/js-for-compiler#types)).